Library/Homebrew/cmd/upgrade.rb

```diff
@@ -171,4 +181,167 @@ def upgrade_formula(f)
+    end
+
+    [formulae_to_upgrade, formulae_pinned]
+  end
+
+  def broken_dependents(kegs, formulae)
+    formulae_to_reinstall = Set.new
+    formulae_pinned_and_outdated = Set.new
+
+    CacheStoreDatabase.use(:linkage) do |db|
+      formulae.each do |formula|
+        descendants = Set.new
+
+        dependents = kegs.select do |keg|
+          keg.runtime_dependencies
+             .any? { |d| d["full_name"] == formula.full_name }
+        end
+
+        next if dependents.empty?
+
+        dependents.each do |keg|
+          f = keg.to_formula
+
+          next if formulae_to_reinstall.include?(f)
+          next if formulae_pinned_and_outdated.include?(f)
+
+          checker = LinkageChecker.new(keg, cache_db: db)
+
+          if checker.broken_library_linkage?
+            if f.outdated?(fetch_head: ARGV.fetch_head?)
+              # Outdated formulae = pinned formulae (see function above)
+              formulae_pinned_and_outdated << f
+            else
+              formulae_to_reinstall << f
+            end
+          end
+
+          descendants << f
+        end
+
+        descendants_to_reinstall = broken_dependents(kegs, descendants)
+
+        formulae_to_reinstall.merge descendants_to_reinstall
+      end
+    end
+
+    formulae_to_reinstall
+  end
+
+  def check_dependents(formulae)
+    return if formulae.empty?
+
+    # First find all the outdated dependents.
+    kegs = Formula.installed
+                  .map(&:opt_or_installed_prefix_keg)
+                  .reject(&:nil?)
+                  .reject { |f| f.runtime_dependencies.to_a.empty? }
+
+    return if kegs.empty?
+
+    oh1 "Checking dependents for outdated formulae" if ARGV.verbose?
+    upgradable, pinned = upgradable_dependents(kegs, formulae).map(&:to_a)
+
+    # Print the pinned dependents.
+    unless pinned.empty?
+      ohai "Not upgrading #{Formatter.pluralize(pinned.length, "broken, but pinned dependent")}:"
+      puts pinned.map { |f| "#{f.full_specified_name} #{f.pkg_version}" } * ", "
+    end
+
+    # Print the upgradable dependents.
+    if upgradable.empty?
+      ohai "No dependents to upgrade" if ARGV.verbose?
+    else
+      ohai "Upgrading #{Formatter.pluralize(upgradable.length, "dependent")}:"
+      formulae_upgrades = upgradable.map do |f|
+        if f.optlinked?
+          "#{f.full_specified_name} #{Keg.new(f.opt_prefix).version} -> #{f.pkg_version}"
+        else
+          "#{f.full_specified_name} #{f.pkg_version}"
+        end
+      end
+      puts formulae_upgrades.join(", ")
+    end
+
+    upgrade_formulae(upgradable)
+
+    # Assess the dependents tree again.
+    kegs = Formula.installed
+                  .map(&:opt_or_installed_prefix_keg)
+                  .reject(&:nil?)
+                  .reject { |f| f.runtime_dependencies.to_a.empty? }
+
+    oh1 "Checking dependents for broken library links" if ARGV.verbose?
+    reinstallable, pinned = broken_dependents(kegs, formulae).map(&:to_a)
```

Comment:
I think `broken_dependents` is not returning anything about pinned formulae; it's just returning `formulae_to_reinstall`

also, I get the following error at this line:

~~~
Error: undefined method `to_a' for #<Formulary::FormulaNamespace85f259522d1d32129f89b1cc281316cc::IgnitionMsgs1:0x0000000106016538>
Did you mean?  to_hash
               to_s
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/set.rb:306:in `each_key'
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/set.rb:306:in `each'
/usr/local/Homebrew/Library/Homebrew/cmd/upgrade.rb:313:in `map'
/usr/local/Homebrew/Library/Homebrew/cmd/upgrade.rb:313:in `check_dependents'
/usr/local/Homebrew/Library/Homebrew/cmd/upgrade.rb:86:in `upgrade'
/usr/local/Homebrew/Library/Homebrew/brew.rb:89:in `<main>'
~~~

Reply:
Those errors are one and the same; `broken_dependents` is supposed to return an `Array` (`.map`) containing two `Set`s (`&:to_a`) of `Formula`e. Fixed now.